text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/resolver/Formatter.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * This file is part of adventure, licensed under the MIT License.
+ *
+ * Copyright (c) 2017-2022 KyoriPowered
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.minimessage.tag.resolver;
+
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
+import java.text.NumberFormat;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAccessor;
+import java.util.Locale;
+import net.kyori.adventure.text.minimessage.tag.Tag;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Tag resolvers producing tags that insert formatted values.
+ *
+ * <p>These are effectively placeholders.</p>
+ *
+ * @since 4.10.0
+ */
+public final class Formatter {
+  private Formatter() {
+  }
+
+  /**
+   * Creates a replacement that inserts a number as a component. The component will be formatted by the provided DecimalFormat.
+   *
+   * <p>This tag expects a format as attribute. Refer to {@link DecimalFormat} for usable patterns.</p>
+   *
+   * <p>You can specify the decimal and grouping separator by adding those as another attribute.</p>
+   *
+   * <p>This replacement is auto-closing, so its style will not influence the style of following components.</p>
+   *
+   * @param key the key
+   * @param number the number
+   * @return the placeholder
+   * @since 4.10.0
+   */
+  public static TagResolver formatNumber(final @NotNull String key, final Number number) {
```

Comment:
Needs some nullability annotations (and on the other method too). Also, perhaps overloads for primitives? Might be a bit boring and annoying tbh.

Reply:
I added the @NotNull annotations. I'm not sure whether we should add the different number primitives.

Reply:
Need them on the return types too! Yeah, idk about different primitives - maybe someone else can weigh in, @zml2008?

Reply:
I'm not sure it's necessary? Number is prooobably fine?